functions/slack/main.py

```diff
@@ -26,13 +28,24 @@
 kgsearch = apiclient.discovery.build(
     'kgsearch',
     'v1',
-    developerKey=os.environ['API_KEY'] or config['KG_API_KEY'])
+    developerKey=os.environ.get('API_KEY') or config.get('KG_API_KEY'),
+    cache_discovery=False)
 # [END functions_slack_setup]
 
 
 # [START functions_verify_webhook]
-def verify_web_hook(form):
-    if not form or form.get('token') != config['SLACK_TOKEN']:
+# Python 3+ version of https://github.com/slackapi/python-slack-events-api/blob/master/slackeventsapi/server.py
+def verify_signature(request):
+    timestamp = request.headers.get('X-Slack-Request-Timestamp', '')
+    signature = request.headers.get('X-Slack-Signature', '')
+
+    req = str.encode('v0:' + str(timestamp) + ':') + request.get_data()
+    request_hash = 'v0=' + hmac.new(
```

Comment:
@crwilcox any ideas on other ways to do this? (Maybe a format string, or moving the concat to line 48?)

Reply:
I think `request.headers` should already be a `dict(str:str)` so the cast doesn't seem necessary. I think format strings might also be preferable to concatenation as well?

The encode could also be moved to the `timestamp =` call but probably doesn't matter much?

```
    req = 'v0:{}:{}'.format(str.encode(timestamp), request.get_data())
    hash = hmac.new(
        str.encode(config['SLACK_SECRET']),
        req, hashlib.sha256
    ).hexdigest()
    request_hash = 'v0={}'.format(hash)
```